### ISSUE TITLE
Parallel ForEach Proposal

### ISSUE THREAD

**Issue by cleitonmarx** [ahmetalpbalkan/go-linq#58]
@ahmetalpbalkan, I think the ForEach function can be an awesome feature for the library. I personally have some interesting use cases in my work that can be better expressed using the ForEach approach.

This code is another try to implement the @kalaninja idea:
```go
func (q Query) ForEach(c chan interface{}, action func(interface{})) {
	next := q.Iterate()
	if c == nil {
		for item, ok := next(); ok; item, ok = next() {
			action(item)
		}
	} else {
		var wg sync.WaitGroup
		for item, ok := next(); ok; item, ok = next() {
			wg.Add(1)
			go func(wg *sync.WaitGroup, item interface{}) {
				defer wg.Done()
				action(item)
			}(&wg, item)
		}

		go func() {
			wg.Wait()
			close(c)
		}()
	}
}
```

This provides an elegant way to solve the concurrent example problem:
```go
results := make(chan interface{})
domains := []string{"www.google.com", "www.github.com", "www.facebook.com"}

From(domains).
    Select(func(d interface{}) interface{} {
        return fmt.Sprintf("http://%s/favicon.ico", d)
    }).
    ForEach(results, func(url interface{}) {
	  resp, _ := http.Get(url.(string))
	  results <- resp.StatusCode
    })

count := FromChannel(results).CountWith(func(r interface{}) bool {
    return r.(int) == 200
})

fmt.Println("Count:", count)
```

Or just for run simple tasks in parallel:
```go
From(domains).
    Select(func(d interface{}) interface{} {
	return fmt.Sprintf("http://%s/favicon.ico", d)
    }).
    ForEach(results, func(url interface{}) {
	resp, _ := http.Get(url.(string))
	fmt.Println(url.(string), "->", resp.StatusCode)
    })

<-results

fmt.Println("Finished all tasks")
```

If no channel was provided, ForEach will iterate sequentially as proposed by @kalaninja.

**Comment by kalaninja** [comment]
it is a good example, though I think that adding extra logic to the method will make it not so lightweight as we try to keep it. I suggest to implement this logic in plinq api, which is probable to happen sometime in the future.

**Comment by ahmetalpbalkan** [comment]
@kalaninja Agreed. Normally, we could have `ForEach` and `ForEachAsParallel`, but I bet most people will get the job done with a simple `go f()` inside ForEach + `sync.WaitGroup`

Personally, if I have a slice of 100,000 items, I don‘t want ForEach to create a goroutine for each of them in the background (this is an easy way to run out of file descriptors on linux and crash your program).

I feel _most_ Go programmers want to have finer control over their concurrency code. For instance instead of running 100,000 goroutines concurrently, I can develop a system that would control max concurrent jobs etc. I think I'd like to see a simpler design which we can incorporate, before tagging v3.0.

As per the design I think I am fine with @kalaninja's original design which was:

- `func (Query) ForEach(func(v interface{})`
- `func (Query) ForEachIndexed(i int, func(v interface{})`
- `ForEachT`, `ForEachIndexedT`

Thoughts?

**Comment by ahmetalpbalkan** [comment]
Closing this as #57 is merged.

@kalaninja If your commit message (not the title) or the PR descriptionincludes `Closes #58`, such issues will be automatically closed. More info:

https://help.github.com/articles/closing-issues-via-commit-messages/
https://github.com/blog/1506-closing-issues-via-pull-requests

**Comment by kalaninja** [comment]
thanks @ahmetalpbalkan. I always forget about this feature when commiting changes
